openai_api/openai_api.py

```diff
@@ -0,0 +1,646 @@
+# Requirement:
+#   pip install "openai<1.0"
+# Usage:
+#   python openai_api_issue.py
+# Visit http://localhost:8000/docs for documents.
+
+import base64
+import copy
+import json
+import time
+from argparse import ArgumentParser
+from contextlib import asynccontextmanager
+from pprint import pprint
+from typing import Dict, List, Literal, Optional, Union
+
+import torch
+import uvicorn
+from fastapi import FastAPI, HTTPException
+from fastapi.middleware.cors import CORSMiddleware
+from pydantic import BaseModel, Field
+from sse_starlette.sse import EventSourceResponse
+from starlette.middleware.base import BaseHTTPMiddleware
+from starlette.requests import Request
+from starlette.responses import Response
+from transformers import AutoModelForCausalLM, AutoTokenizer, Qwen2ForCausalLM, Qwen2Tokenizer
+from transformers.generation import GenerationConfig
+from qwen2chat import Qwen2ForChatLM
```

Comment:
where??? I can not find.

Reply:
In line 471:
response, _ = model.chat(
            tokenizer,
            query,
            history=history,
            system=system,
            stop_words_ids=stop_words_ids,
            **gen_kwargs,
        )

The chat interface supports function calling now.